paddle/fluid/extension/include/op_function.h

```diff
@@ -0,0 +1,296 @@
+
+// Record forward function traits
+using FuncInfo = std::pair<size_t, size_t>;
+using TraitsFunc = FuncInfo (*)();
+
+template <typename T, typename Enabled = void>
+struct KernelFuncTraits;
+
+template <typename Return, typename... Args>
+struct KernelFuncTraits<Return (*)(Args...)> {
+  static FuncInfo GetFuncInfo() {
+    // TODO(chenweihang): parse tensor args num & attribute num
+    // Now only Tensor is input by default
+    return std::make_pair(Arity, 0);
+  }
+
+ private:
+  using ReturnType = Return;
+  using ArgsTuple = std::tuple<Args...>;
+
+  enum : std::size_t { Arity = sizeof...(Args) };
+
+  template <std::size_t Index>
+  using Arg = typename std::tuple_element<Index, ArgsTuple>::type;
+};
+
+#define OP_INFO(...) \
+  ::paddle::KernelFuncTraits<decltype(&__VA_ARGS__)>::GetFuncInfo
+
+////////////////////// Kernel Function (PD_KERNEL) ////////////////////////
+
+// Record Op kernel core function
+using KernelFunc = std::vector<Tensor> (*)(std::vector<const Tensor*> inputs,
+                                           std::vector<boost::any> attrs);
+
+template <typename T>
+struct TypeTag {};
+
+template <typename F, F f>
+struct KernelFuncImpl;
+
+template <typename Return, typename... Args, Return (*impl_fn)(Args...)>
+struct KernelFuncImpl<Return (*)(Args...), impl_fn> {
+  static Return Compute(std::vector<const Tensor*> inputs,
+                        std::vector<boost::any> attrs) {
+    return ComputeCallHelper<Args..., TypeTag<int>>::template Compute<0, 0>(
+        inputs, attrs);
+  }
+
+ private:
+  template <typename... RemainingArgs>
+  struct ComputeCallHelper;
+
+  // for Tensor input
+  template <typename... Tail>
+  struct ComputeCallHelper<const Tensor&, Tail...> {
+    template <int in_idx, int attr_idx, typename... PreviousArgs>
+    static Return Compute(std::vector<const Tensor*> inputs,
+                          std::vector<boost::any> attrs,
+                          const PreviousArgs&... pargs) {
+      static_assert(attr_idx == 0,
+                    "Input tensor should appear before attributes.");
+      const Tensor& arg = *(inputs[in_idx]);
+      return ComputeCallHelper<Tail...>::template Compute<in_idx + 1, attr_idx>(
+          inputs, attrs, pargs..., arg);
+    }
+  };
+
+  // for int attribute input (not used now)
+  template <typename... Tail>
+  struct ComputeCallHelper<int, Tail...> {
+    template <int in_idx, int attr_idx, typename... PreviousArgs>
+    static Return Compute(std::vector<const Tensor*> inputs,
+                          std::vector<boost::any> attrs,
+                          const PreviousArgs&... pargs) {
+      try {
+        int arg = boost::any_cast<int>(attrs[attr_idx]);
+        return ComputeCallHelper<Tail...>::template Compute<in_idx,
+                                                            attr_idx + 1>(
+            inputs, attrs, pargs..., arg);
+      } catch (boost::bad_any_cast&) {
+        std::ostringstream err;
```

Comment:
change to remove reference on data_type.h

Reply:
done, thx